src/main/kotlin/org/rust/ide/inspections/RustDeprecatedInspection.kt

```diff
@@ -0,0 +1,106 @@
+            override fun visitStructExprField(el: RustStructExprFieldElement) = inspectReference(el.reference, el.identifier, holder)
+            override fun visitTraitRef(el: RustTraitRefElement) = inspectReference(el.path.reference, el, holder)
+            override fun visitUseGlob(el: RustUseGlobElement) = inspectReference(el.reference, el, holder)
+            override fun visitUseItem(el: RustUseItemElement) {
+                if (el.useGlobList == null) {
+                    inspectReference(el.path?.reference, el.path, holder)
+                }
+            }
+        }
+
+    fun inspectReference(ref: RustReference?, el: PsiElement?, holder: ProblemsHolder) {
+        if (ref == null || el == null) return
+        val deprInfo = ref.resolve()?.findDeprecation() ?: return
+        var msg = "${deprInfo.element.getTitle()} is deprecated"
+        if (deprInfo.since != null) {
+            msg += " since ${deprInfo.since}"
+        }
+        if (deprInfo.reason != null) {
+            msg += ": ${deprInfo.reason}"
+        }
+        if (deprInfo.note != null) {
+            msg += ": ${deprInfo.note}"
+        }
+        holder.registerProblem(el, msg, ProblemHighlightType.LIKE_DEPRECATED)
+    }
+
+    /**
+     * Detects if element is deprecated and returns the title of the exact tree node
+     * that causes deprecation.
+     */
+    private fun PsiElement.findDeprecation(): DeprecationInfo? {
+        var attr: RustAttrElement? = null
+        if (this is RustOuterAttributeOwner) {
+            attr = findOuterAttr("deprecated") ?: findOuterAttr("rustc_deprecated")
+        }
+        if (attr == null && this is RustInnerAttributeOwner) {
+            attr = findInnerAttr("deprecated") ?: findInnerAttr("rustc_deprecated")
+        }
+        if (attr == null && parent != null) {
+            return parent.findDeprecation()
+        }
+        return if (attr != null) DeprecationInfo(this, attr) else null
+    }
+
+    private fun PsiElement.getTitle(): String = when (this) {
+        is RustConstItemElement -> "Constant `$name`"
+        is RustEnumItemElement -> "Enum `$name`"
+        is RustEnumVariantElement -> "Enum variant `$name`"
+        is RustFnItemElement -> "Function `$name`"
+        is RustFieldDeclElement -> "Field `$name`"
+        is RustImplMethodMemberElement -> "Method `$name`"
+        is RustModItemElement -> "Module `$name`"
+        is RustStaticItemElement -> "Static constant `$name`"
+        is RustStructItemElement -> "Type `$name`"
+        is RustTraitItemElement -> "Trait `$name`"
+        is RustTypeItemElement -> "Type `$name`"
+        is RustNamedElement -> "`$name`"
+        else -> "Item"
+    }
+}
+
+/**
+ * Represents information about deprecated element.
+ */
+private class DeprecationInfo(
+    val element: PsiElement,
+    attr: RustAttrElement
+) {
+    var since: String? = null
+    var note: String? = null
+    var reason: String? = null
```

Comment:
Can we use `val`s here? 

You'll need three local variables for this in `init`, which is a bit more code, but immutability is a great thing!